common/OpTestSSH.py

```diff
@@ -155,52 +159,133 @@ def get_console(self):
+          matching = [xs for xs in sudo_responses if any(xs in xa for xa in console.after.splitlines())]
+          if len(matching):
+            # remove the expect prompt since matched generic #
+            del handle_list_output[-1]
+            echo_rc = 1
+            rc = -1 # use this to flag the failure next
+        if rc == 0:
+          # with unknown prompts and unknown environment unable to capture echo $?
+          echo_rc = 0
+          self.set_env(console)
+          list_output = handle_list_output
+        elif rc == 1:
+          retry_list_output, echo_rc = self.retry_password(console, command)
+          list_output = (handle_list_output + retry_list_output)
+        else:
+          if (rc == 2) or (rc == 3):
+            failure_list_output += ['Password Problem/TIMEOUT ']
+            failure_list_output += pre_combo_io.splitlines()
+          # timeout path needs access to output
+          # handle_list_output empty if timeout or EOF
+          failure_list_output += handle_list_output
+          raise CommandFailed(command, ''.join(failure_list_output), -1)
+        return list_output, echo_rc
+
+    def run_command(self, command, timeout=60):
+        running_sudo_s = False
+        extra_sudo_output = False
+        console = self.get_console()
+        expect_prompt = self.build_prompt() + "$"
+        console.sendline(command)
+        console.expect("\n") # removes the echo of command from output
+        if command == 'sudo -s':
+          running_sudo_s = True
+          # special case to catch loss of env
+          rc = console.expect([".*#", r"[Pp]assword", pexpect.TIMEOUT, pexpect.EOF], timeout=timeout)
+        else:
+          rc = console.expect([expect_prompt, r"[Pp]assword", pexpect.TIMEOUT, pexpect.EOF], timeout=timeout)
+        output_list = []
+        output_list += console.before.splitlines()
+        # if we are running 'sudo -s' as root then catch on generic # prompt, restore env
+        if running_sudo_s and (rc == 0):
+          extra_sudo_output = True
+          set_env_list = self.set_env(console)
         if rc == 0:
-            res = output
-            res = res.splitlines()
-            if exitcode != 0:
-                raise CommandFailed(command, res, exitcode)
-            return res
+          if extra_sudo_output:
+            output_list += set_env_list
+          console.sendline("echo $?")
+          rc2 = console.expect([expect_prompt, pexpect.TIMEOUT, pexpect.EOF], timeout=timeout)
+          if rc2 == 0:
+            echo_output = console.before
+            echo_rc = int(echo_output.splitlines()[-1])
+          else:
+            raise CommandFailed(command, 'echo TIMEOUT', -1)
+        elif rc == 1:
+          handle_output_list, echo_rc = self.handle_password(console, command)
+          # remove the expect prompt since matched generic #
+          del handle_output_list[-1]
+          output_list = handle_output_list
         else:
```

Comment:
Here we are sending ctrl-c for both pexpect.TIMEOUT and pexpect.EOF. I think this should handle for both separately. If timeout occurs send cntrl-c is fine. If EOF occurs means set the state to Disconnected. so that next run_command can re-launch the console/ssh session. After that raise the "Session EOF" exception.
```
        self.state = ConsoleState.DISCONNECTED
```

Reply:
Fixed